validator/parse-css.js

```diff
@@ -827,104 +350,341 @@ Canonicalizer.prototype.blockTypeFor = function(atRule) {
 };
 
 /**
- * @param {!parse_css.Stylesheet} stylesheet
+ * Parses and returns a list of rules, such as at the top level of a stylesheet.
+ * Return list has only QualifiedRule's and AtRule's as top level elements.
+ * @param {!Array<!parse_css.CSSParserToken>} tokenList
+ * @param {boolean} topLevel
  * @param {!Array<!parse_css.ErrorToken>} errors output array for the errors.
- * @return {!parse_css.Stylesheet}
+ * @return {!Array<!parse_css.CSSParserRule>}
  */
-Canonicalizer.prototype.canonicalizeStylesheet = function(stylesheet, errors) {
-  for (let i = 0; i < stylesheet.value.length; i++) {
-    const currentRule = stylesheet.value[i];
-    if (currentRule instanceof parse_css.EOFToken) {
-      stylesheet.eof = currentRule;
-    } else if (currentRule instanceof parse_css.QualifiedRule) {
-      stylesheet.rules.push(
-          this.canonicalizeQualifiedRule(currentRule, errors));
-    } else if (currentRule instanceof parse_css.AtRule) {
-      stylesheet.rules.push(this.canonicalizeAtRule(currentRule, errors));
+Canonicalizer.prototype.parseAListOfRules = function(
+    tokenList, topLevel, errors) {
+  const tokenStream = new parse_css.TokenStream(tokenList);
+  const rules = [];
+  while (true) {
+    tokenStream.consume();
+    if (tokenStream.current() instanceof parse_css.WhitespaceToken) {
+      continue;
+    } else if (tokenStream.current() instanceof parse_css.EOFToken) {
+      return rules;
+    } else if (tokenStream.current() instanceof parse_css.CDOToken ||
+               tokenStream.current() instanceof parse_css.CDCToken) {
+      if (topLevel) {
+        continue;
+      }
+      this.parseAQualifiedRule(tokenStream, rules, errors);
+    } else if (tokenStream.current() instanceof parse_css.AtKeywordToken) {
+      rules.push(this.parseAnAtRule(tokenStream, errors));
     } else {
-      goog.asserts.fail('unexpected rule inside a top-level stylesheet: ' +
-          currentRule.name);
+      this.parseAQualifiedRule(tokenStream, rules, errors);
     }
   }
+}
 
-  // Clear the value field, just to reduce mess in outputs.
-  stylesheet.value = [];
+/**
+ * Parses an At Rule.
+ *
+ * @param {!parse_css.TokenStream} tokenStream
+ * @param {!Array<!parse_css.ErrorToken>} errors output array for the errors.
+ * @return {!parse_css.AtRule}
+ */
+Canonicalizer.prototype.parseAnAtRule = function(tokenStream, errors) {
+  if (!(tokenStream.current() instanceof parse_css.AtKeywordToken))
+    throw 'Internal Error: parseAnAtRule precondition not met';
```

Comment:
All `throw` should be `throw new Error(...)`. In this case, ideally, it'd be an `assert`